src/cxx_supportlib/vendor-copy/adhoc_lve.h

```diff
@@ -0,0 +1,297 @@
+#else
+#	define LIBLVE_LOAD(fname) \
+		fname ## _function_ptr = \
+				(fname ## _function_ptr_t) dlsym(liblve_handle.get(), #fname); \
+		if (!fname ## _function_ptr && err_buf.str().empty()) \
+			init_error = (std::string) "Failed to init LVE library " + ::dlerror()
+
+	LibLve()
+	  : init_lve_function_ptr        (NULL)
+	  , destroy_lve_function_ptr     (NULL)
+	  , lve_enter_flags_function_ptr (NULL)
+	  , lve_exit_function_ptr        (NULL)
+	{
+		void *handle = ::dlopen("liblve.so.0", RTLD_LAZY);
+		if (handle) {
+			liblve_handle.reset(handle, ::dlclose);
+		} else {
+			// No liblve found, but it is OK and we are running
+			// on non LVE capable system
+			return;
+		}
+
+		std::ostringstream err_buf;
+
+		LIBLVE_LOAD(init_lve);
+		LIBLVE_LOAD(destroy_lve);
+		LIBLVE_LOAD(lve_enter_flags);
+		LIBLVE_LOAD(lve_exit);
+		LIBLVE_LOAD(jail);
+
+		if (!err_buf.str().empty())
+		{
+			init_error = err_buf.str();
+			return;
+		}
+
+		struct liblve* init_handle = init_lve_function_ptr(std::malloc, std::free);
+		if (!init_handle)
+			err_buf << "init_lve error [" << errno << "]";
+		else
+			lve_init_handle.reset(init_handle, destroy_lve_function_ptr);
+
+		init_error = err_buf.str();
+	}
+
+#undef LIBLVE_LOAD
+#endif
+
+	bool is_error() const { return !init_error.empty(); }
+	const std::string& error() const { return init_error; }
+
+	bool is_lve_available() const { return (bool) liblve_handle; }
+	bool is_lve_ready() const { return is_lve_available() && !is_error(); }
+
+	int jail(const struct passwd *pw, std::string& jail_err) const
+	{
+		char error_msg[8192];
+		int rc = jail_function_ptr(pw, error_msg);
+		if (rc < 0)
+			jail_err.assign(error_msg);
+		return rc;
+	}
+
+private:
+	friend class LveEnter;
+
+	/**
+	 * using boost::shared_ptr because of scoped_ptr does not allow
+	 * to specify custom destructor
+	 */
+	boost::shared_ptr<void>          liblve_handle;
+	boost::shared_ptr<struct liblve> lve_init_handle;
+
+	init_lve_function_ptr_t          init_lve_function_ptr;
+	destroy_lve_function_ptr_t       destroy_lve_function_ptr;
+	lve_enter_flags_function_ptr_t   lve_enter_flags_function_ptr;
+	lve_exit_function_ptr_t          lve_exit_function_ptr;
+	jail_function_ptr_t              jail_function_ptr;
+
+	std::string                      init_error;
+};
+
+class LveInitSignleton {
+public:
+	static LibLve& getInstance(std::string* outInitOneTimeError) {
+		if (!instance())
+		{
+			static boost::once_flag flag = BOOST_ONCE_INIT;
+			boost::call_once(flag, initOnce, outInitOneTimeError);
+		}
+
+		return *instance();
+	}
+private:
+	typedef boost::scoped_ptr<LibLve> lvelib_scoped_ptr;
+
+	static lvelib_scoped_ptr& instance() {
+		static lvelib_scoped_ptr lvelib_scoped_ptr;
+		return lvelib_scoped_ptr;
+	}
+
+	static void initOnce(std::string *err_buf) {
+		instance().reset(new LibLve);
+
+		if (instance()->is_error() && err_buf)
+			*err_buf = instance()->error();
+	}
+};
+
+class LveEnter : private boost::noncopyable {
```

Comment:
Please do not put a space before the `:`:

    class LveEnter: private boost::noncopyable {